File: cpp/linked_list.cpp
#include <memory>
#include <vector>
#include <iostream>

struct node {
    int value;
    std::unique_ptr<node> next;
    node(int value) : value{value}, next{nullptr}
    {

    } 
};

struct linked_list{
    std::unique_ptr<node> head;
    node* tail;
    size_t size; 

    linked_list()
        : head{nullptr}, tail{nullptr} ,size{0}
    {
    }

    void push_back(int value) {
        auto new_tail = std::make_unique<node>(value);
        auto raw_tail = new_tail.get();
        if (this->tail != nullptr) {
            tail->next = std::move(new_tail);
        } else {
            this->head = std::move(new_tail);
        }
        this->tail = raw_tail;
        this->size++;
    }

    int remove_head() {
        if (this->head) {
            auto old_value = this->head->value;

            auto new_head = std::move(this->head->next);
            if (new_head == nullptr) {
                this->tail = nullptr;
            }
            this->head = std::move(new_head);
            this->size--;
            return old_value;
        }        
    }

    void print() {
        auto node = this->head.get();
        while (node) {
            std::cout << node->value << ", ";
            node = node->next.get();
        }
        std::cout << std::endl;
    }
};

template<typename ADT, typename Pred>
void generic_print(ADT ds, Pred pred) {
    for(auto iter = ds.begin(); iter != ds.end(); iter++){
        pred(*iter);
    }
}

int main() {
    auto l = linked_list();
    auto arr = std::vector<int>();
    for (auto i = 0; i < 10; i++) {
        l.push_back(i);
        arr.push_back(i);
    }
    l.print();
    while (l.size > 0) {
        l.remove_head();
        l.print();
    }
    l.print();

    std::cout << "functors >>>\n";
    generic_print(arr, [](auto x) {
        std::cout << x << ", ";
    });
    std::cout << "<<<\n";
    return 0;
}

File: rust/src/linked_list.rs
struct Node {
    value: i32,
    next: Option<Box<Node>> 
}

struct LinkedList {
    head: Option<Box<Node>>,
    tail: Option<*mut Node>
}

impl LinkedList {
    fn new() -> Self {
        LinkedList{
            head: None,
            tail: None
        }
    }
    fn push_back(&mut self, value: i32) {
        let mut new_tail = Box::new(Node { value, next: None });
        let raw_tail: *mut _ = &mut *new_tail;

        if self.tail.is_some() {
            unsafe { (*self.tail.unwrap()).next = Some(new_tail) };
        } else {
            self.head = Some(new_tail);
        }
        self.tail = Some(raw_tail);
    }

    pub fn pop_back(&mut self) -> Option<i32> {
        if let Some(head) = &mut self.head {
            let old_value = Some(head.value);
            let new_head = head.next.take();
            if new_head.is_none() {
                self.tail = None;
            };
            self.head = new_head;
            old_value
        } else {
            None
        }
    }
    fn print(&mut self) {
        let mut node  = &self.head;
        while let Some(old_node) = node  {
            match &node {
                Some(node) => print!("{}", node.value),
                _ => () ,
            }
            node = &old_node.next;
        }
        println!("")
    }
}

fn main() {
    let mut list = LinkedList::new();
    list.push_back(0);
    list.push_back(1);
    list.push_back(2);
    list.push_back(3);
    list.print();
}